pkg/cloudprovider/aws/aws.go

```diff
@@ -489,3 +808,338 @@ func (self *AWSCloud) GetZone() (cloudprovider.Zone, error) {
+	glog.V(2).Infof("CreateTCPLoadBalancer(%v, %v, %v, %v, %v)", name, region, externalIP, port, hosts)
+
+	if affinity != api.AffinityTypeNone {
+		// ELB supports sticky sessions, but only when configured for HTTP/HTTPS
+		return "", fmt.Errorf("unsupported load balancer affinity: %v", affinity)
+	}
+
+	if len(externalIP) > 0 {
+		return "", fmt.Errorf("External IP cannot be specified for AWS ELB")
+	}
+
+	instances, err := self.getInstancesByDnsNames(hosts)
+	if err != nil {
+		return "", err
+	}
+
+	vpc, err := self.findVpc()
+	if err != nil {
+		glog.Error("error finding vpc", err)
+		return "", err
+	}
+	if vpc == nil {
+		return "", fmt.Errorf("Unable to find vpc")
+	}
+
+	// Construct list of configured subnets
+	subnetIds := []string{}
+	{
+		subnets, err := self.ec2.DescribeSubnets(nil, vpc.VpcId)
+		if err != nil {
+			return "", err
+		}
+
+		//	zones := []string{}
+		for _, subnet := range subnets {
+			subnetIds = append(subnetIds, subnet.SubnetId)
+			if !strings.HasPrefix(subnet.AvailabilityZone, region) {
+				glog.Error("found AZ that did not match region", subnet.AvailabilityZone, " vs ", region)
+				return "", fmt.Errorf("invalid AZ for region")
+			}
+			//		zones = append(zones, subnet.AvailabilityZone)
+		}
+	}
+
+	// Build the load balancer itself
+	var loadBalancerName, dnsName string
+	{
+		loadBalancer, err := self.describeLoadBalancer(region, name)
+		if err != nil {
+			return "", err
+		}
+
+		if loadBalancer == nil {
+			createRequest := &elb.CreateLoadBalancer{}
+			// TODO: Is there a k8s UUID that it would make sense to use?
+			uuid := strings.Replace(string(util.NewUUID()), "-", "", -1)
+			awsId := LOADBALANCER_NAME_PREFIX + uuid
+			if len(awsId) > LOADBALANCER_NAME_MAXLEN {
+				awsId = awsId[:LOADBALANCER_NAME_MAXLEN]
+			}
+			createRequest.LoadBalancerName = awsId
+
+			listener := elb.Listener{}
+			listener.InstancePort = int64(port)
+			listener.LoadBalancerPort = int64(port)
+			listener.Protocol = "tcp"
+			listener.InstanceProtocol = "tcp"
+			createRequest.Listeners = []elb.Listener{listener}
+
+			// TODO: Should we use a better identifier (the kubernetes uuid?)
+			//	nameTag := &elb.Tag{ Key: "Name", Value: name}
+			//	createRequest.Tags = []Tag { nameTag }
+
+			//	zones := []string{"us-east-1a"}
+			//	createRequest.AvailZone = removeDuplicates(zones)
+
+			// We are supposed to specify one subnet per AZ.
+			// TODO: What happens if we have more than one subnet per AZ?
+			createRequest.Subnets = subnetIds
+
+			sgName := "k8s-elb-" + name
+			sgDescription := "Security group for Kubernetes ELB " + name
+
+			{
+				// TODO: Should we do something more reliable ?? .Where("tag:kubernetes-id", kubernetesId)
+				securityGroups, err := self.ec2.DescribeSecurityGroups(nil, sgName, vpc.VpcId)
+				if err != nil {
+					return "", err
+				}
+				var securityGroupId string
+				for _, securityGroup := range securityGroups {
+					securityGroupId = securityGroup.Id
+				}
```

Comment:
This is an interesting way to get the ID of the last security group. Any reason why you're doing it this way?